cli/launchpad/folder.go

```diff
@@ -0,0 +1,145 @@
+package launchpad
+
+import (
+	"fmt"
+	"log"
+	"strings"
+)
+
+const (
+	folderNameMin = 3
+	folderNameMax = 30
+)
+
+// folderSpecYAML defines GCP Folder's spec.
+type folderSpecYAML struct { // Inner mappings
+	Id             string                 `yaml:"id"`
+	DisplayName    string                 `yaml:"displayName"`
+	ParentRef      referenceYAML          `yaml:"parentRef"`
+	SubFolderSpecs []*folderSpecYAML      `yaml:"folders"`
+	Undefined      map[string]interface{} `yaml:",inline"` // Catch-all for untended behavior
+}
+
+// folders represents a list of folders.
+type folders []*folderYAML
+
+// add appends a folder into the folder list if it does not exist already.
+func (fs *folders) add(newF *folderYAML) {
+	for _, f := range *fs {
+		if f.Spec.Id == newF.Spec.Id { // silently ignore already existing folder
+			return
+		}
+	}
+	*fs = append(*fs, newF)
+}
+
+// folderYAML is a GCP Folder YAML representation.
+type folderYAML struct {
+	headerYAML `yaml:",inline"`
+	Spec       folderSpecYAML `yaml:"spec"`
+	subFolders folders        // subFolder represents validated sub directories.
+}
+
+// refId returns an internal referencable id.
+func (f *folderYAML) refId() string { return fmt.Sprintf("%s.%s", Folder, f.Spec.Id) }
+
+// String implements Stringer and generates a string representation.
+func (f *folderYAML) String() string { return strings.Join(f.dump(0), "\n") }
+
+// validate ensures input YAML fields are correct.
+//
+// validate also populates subFolders.
+func (f *folderYAML) validate() error {
+	if f.Spec.Id == "" {
+		return errMissingRequiredField
+	}
+	switch f.Spec.ParentRef.TargetTypeStr { // Validate Supported Parents
+	case Organization.String(), Folder.String():
+	default:
+		log.Printf("fatal: unsupported parent '%s' type for Folder\n", f.Spec.ParentRef.TargetTypeStr)
+		return errInvalidParent
+	}
+
+	// TODO validate misc
+	if ind := tfNameRegex.FindStringIndex(f.Spec.Id); ind == nil {
+		log.Printf("GCP Folder [%s] ID does not conform to Terraform standard", f.Spec.DisplayName)
+		return errValidationFailed
+	}
+
+	if len(f.Spec.DisplayName) < folderNameMin || len(f.Spec.DisplayName) > folderNameMax {
```

Comment:
Technically length limits could be included in regex (just FYI).

Reply:
This is checking a different field, regex check for id, this check for DisplayName.